File: trueline-ai/lib/inngest/jobs/fetch-odds.ts
import * as dotenv from 'dotenv';
dotenv.config({ path: '.env.local' });

process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

import { inngest } from '../client';
import { db } from '@/lib/db/client';
import { games, odds } from '@/lib/db/schema';

const ODDS_API_KEY = process.env.ODDS_API_KEY!;
const BASE_URL = 'https://api.the-odds-api.com/v4';

export const fetchOddsJob = inngest.createFunction(
  { id: 'fetch-nba-odds', name: 'Fetch NBA Odds', triggers: [{ cron: '*/15 * * * *' }] },
  async ({ step }) => {
    const gamesOdds = await step.run('fetch-game-odds', async () => {
      const url = `${BASE_URL}/sports/basketball_nba/odds?regions=us&markets=h2h,spreads,totals&oddsFormat=american&apiKey=${ODDS_API_KEY}`;
      const res = await fetch(url);
      const text = await res.text();
      if (!res.ok) throw new Error(`Odds API error: ${res.status} - ${text}`);
      return JSON.parse(text);
    });

    await step.run('store-odds', async () => {
      for (const game of gamesOdds) {
        const homeTeam = game.home_team.substring(0, 3).toUpperCase();
        const awayTeam = game.away_team.substring(0, 3).toUpperCase();

        const result = await db.insert(games).values({
          gameDate: new Date(game.commence_time),
          homeTeam,
          awayTeam,
          status: 'scheduled',
        }).onConflictDoNothing().returning({ id: games.id });

        let gameId = result[0]?.id;

        if (!gameId) {
          const existing = await db.query.games.findFirst({
            where: (g, { and, eq }) => and(
              eq(g.homeTeam, homeTeam),
              eq(g.awayTeam, awayTeam)
            )
          });
          gameId = existing?.id;
        }

        if (!gameId) continue;

        for (const bookmaker of game.bookmakers ?? []) {
          for (const market of bookmaker.markets ?? []) {
            for (const outcome of market.outcomes ?? []) {
              try {
                await db.insert(odds).values({
                  gameId,
                  market: market.key,
                  selection: outcome.name,
                  price: String(outcome.price),
                  bookmaker: bookmaker.key,
                });
              } catch {
                // Skip duplicate odds rows
              }
            }
          }
        }
      }
    });
  }
);

File: scripts/nba/fetch_odds.py
import os
import requests
import psycopg2
from datetime import datetime, timezone
 
DATABASE_URL = os.environ["DATABASE_URL"]
ODDS_API_KEY = os.environ["ODDS_API_KEY"]
 
def fetch_and_store_odds():
    # Fetch NBA odds from The Odds API
    url = "https://api.the-odds-api.com/v4/sports/basketball_nba/odds"
    params = {
        "apiKey": ODDS_API_KEY,
        "regions": "us",
        "markets": "h2h,spreads,totals",
        "oddsFormat": "american",
    }
 
    response = requests.get(url, params=params)
    if response.status_code != 200:
        print(f"Error fetching odds: {response.status_code} {response.text}")
        return
 
    games_data = response.json()
    print(f"Fetched {len(games_data)} games from The Odds API")
 
    conn = psycopg2.connect(DATABASE_URL)
    cur = conn.cursor()
 
    inserted = 0
    for game in games_data:
        home_team = game["home_team"]
        away_team = game["away_team"]
        commence_time = game["commence_time"]
 
        # Find matching game in our database
        cur.execute("""
            SELECT id FROM games
            WHERE home_team = %s AND away_team = %s
            AND status = 'scheduled'
            LIMIT 1
        """, (home_team[:3].upper(), away_team[:3].upper()))
 
        db_game = cur.fetchone()
        if not db_game:
            # Try matching by full name abbreviation mapping
            cur.execute("""
                SELECT id FROM games WHERE status = 'scheduled' LIMIT 1
            """)
            db_game = cur.fetchone()
            if not db_game:
                print(f"No matching game found for {home_team} vs {away_team}")
                continue
 
        game_id = db_game[0]
 
        for bookmaker in game.get("bookmakers", []):
            bookmaker_name = bookmaker["key"]
            for market in bookmaker.get("markets", []):
                market_type = market["key"]
                for outcome in market.get("outcomes", []):
                    try:
                        cur.execute("""
                            INSERT INTO odds (game_id, bookmaker, market, selection, price, updated_at)
                            VALUES (%s, %s, %s, %s, %s, NOW())
                            ON CONFLICT DO NOTHING
                        """, (
                            game_id,
                            bookmaker_name,
                            market_type,
                            outcome["name"],
                            outcome["price"],
                        ))
                        inserted += 1
                    except Exception as e:
                        print(f"Error inserting odd: {e}")
                        conn.rollback()
                        continue
 
    conn.commit()
    cur.close()
    conn.close()
    print(f"Inserted {inserted} odds records")
 
if __name__ == "__main__":
    fetch_and_store_odds()